### PR TITLE
Ground Trace from orbit

### PR DESCRIPTION
**Pull request by Charon77**
This PR adds ground trace options in order to figure out where on the surface the vessel will be on.
This is useful for temperature scan missions above certain altitude where we can simply set the orbit such that the orbit crosses the given target.

Picture below demonstrates use case

PS: I really have trouble editing the file due to my unfamiliarity with UTF with BOM. As such, the diffs of the files includes everything, not just the changes that I made. I would be more than happy to help, if only I know how.

image
Image 1: Normal mode

image
Image 2: Dropping ground trace to surface by clicking "Ground Trace mode"

image
Image 3: maneuver node adjusts inclination change to cross target  

image
Image 4: And done!

### CODE REVIEW

**Comment by fat-lobyte** [comment]
The feature looks good, but we really gotta figure out those file changes, haha 😅 

>PS: I really have trouble editing the file due to my unfamiliarity with UTF with BOM. As such, the diffs of the files includes everything, not just the changes that I made. 

First a few questions: what operating system are you using? Which editor are you using? Was it a version of Visual Studio?
If you're on windows, could you please:

1.  download Notepad++
2. open the files that you changed with Notepad++
3. Look in the lower right corner, it should say something like "Windows (CR LF) | UTF-8 BOM". Does it say that or something else?

**Comment by Charon77** [comment]
Alright, I'm using Windows 10, using Visual Studio Community Edition 2015. I have checked with Npp and indeed it says UTF-8 BOM.

I've also tried, from Visual Studio, saving as Unicode with signature (I've read that it's the same as BOM) but still breaks diff.

image

**Comment by fat-lobyte** [comment]
> Alright, I'm using Windows 10, using Visual Studio Community Edition 2015. I have checked with Npp and indeed it says UTF-8 BOM.

That's strange, because when I look at the diff, it seems that you **removed** the BOM from  Plugin/Settings.cs and Plugin/MapGUI.cs

Also, when I check out your files, I get "Windows (CR LF) | UTF-8 " but **no BOM**. But that's not the real problem, I believe.
The real problem here are the line endings, which seem to be CRLF in *both*. I think something is screwy with your Git settings.

What kind of Git are you using (Git for Windows, which version? or GitHub Desktop? Or "GitHub for Windows"?
Could you show me the contents of your .gitconfig? Specifcally, could you run

```
git config core.autocrlf
git config --global core.autocrlf
```

TBH, On one hand I could probably fix up this particular PR to only include non-whitespace changes. On the other hand, I think we should figure out what's wrong to facilitate future contributions (by you and others) ;)

**Comment by Charon77** [comment]
Thank you so much, I really appreciate your help, as I am unfamiliar with this situation.

Here's what happens when I run the commands
`git config core.autocrlf` true
`git config --global core.autocrlf` outputs nothing

I am using Git for Windows on PowerShell 
`git --version` outputs 
`git version 2.8.1.windows.1`

As for whether or not BOM is present: I think I may have converted them to not include the BOM. This is due to my confusion, since simply cloning your repo gets BOM in the beginning of each file, and immediately I got 24+ unstaged changed, even though I didn't do anything. I could try converting the files back to BOM but I still believe something's wrong with my git settings as you mentioned.

Even after deleting my local folder and re-cloning my repo, I immediately got this
`git status`
image

`git diff`
image

P.S so sorry for the terminal color, it's the default and I haven't changed it yet

**Comment by fat-lobyte** [comment]
> I really appreciate your help, as I am unfamiliar with this situation.

That's quite alright, I'd prefer contributers to write code, not deal with whitespace changes ;)

> Even after deleting my local folder and re-cloning my repo, I immediately got this

Yeah, that's not good. This should really not happen, and this needs some fixing on the repository side so that people can just clone and use it, without having everything show up as changed. The frickin line endings have been killing me for a while now.

> I am using Git for Windows on PowerShell

Out of curiosity, could run git status from the freshly checked-out repostitory using Git BASH instead of powershell?

> still believe something's wrong with my git settings as you mentioned.

Could you please copy-paste the contents of .git/config (inside the repository) and .gitconfig (in your users home folder) to a github gist?

**Comment by Charon77** [comment]
Hey, looks like it works!

image

Here's my git configs https://gist.github.com/Charon77/fe8228b0d24af34ffc379664b4865d26

git diff now doesn't return anything, which is great! 😄 

I've updated by git push -f to my branch. I wish there's a more proper way to do it without so many conflicts, though.

Many thanks for your help! Wouldn't have thought it was caused by older version of git

**Comment by fat-lobyte** [comment]
> Hey, looks like it works!

Yay, congrats!

> I've updated by git push -f to my branch. I wish there's a more proper way to do it without so many conflicts, though.

A forced push is actually legit, if it's on a private branch and/or you inform the people who pull from it.
On that note, a little advice: generally, don't work on your master branch. Especially for Pull Requests, it's better to create a named branch just for one feature/bugfix. That's just for the future though, you don't have to rename it now or anything ;)

I'll do some testing and code review when you say you are ready with your feature.

**Comment by fat-lobyte** [comment]
Hey @Charon77 , how is it going? Do you still want to add more commits, or can I start the review process?

**Comment by fat-lobyte** [comment]
Just tested it and it turning the Ground Trace Mode on and off does mess with other parts of the MapOverlay.

In some situations, when Body-Fixed mode is on, toggling Ground Trace mode also toggles the trajectory (red line). It other situations, it makes the targeting cross disappear.

Could you look into that?

**Comment by Charon77** [comment]
@fat-lobyte  Hi,
so sorry for the delay. I was off grid during holiday.

I'll look it up 😄 

Thank you for noticing

**Comment by Charon77** [comment]
@fat-lobyte Hi, I've done some stuffs. I think it's review-worthy now. I've thrown away all the effort I've made to fix the bugs my code introduced and instead got away with a simpler way (and made more sense) that works!

**Comment by fat-lobyte** [comment]
ground-trace

I am getting increasingly confused what you are actually trying to do here. Could you please provide a clear set of instructions (that I can put in the forum post as "manual"), what the result is and what the feature is useful for?

I value your contribution, but reviewing and testing changes takes time for me. Please make sure that your code has high quality and is well tested **before** asking me to merge.
For example, saying "I think it's review-worthy now." when it's clearly buggy and and before pushing yet another commit on top is not a very nice thing to do.

**Comment by Charon77** [comment]
@fat-lobyte  I'm sorry for the confusion caused, as I am new to project contribution.

This feature I'm adding is to allow players to figure out where on the ground the ship will cross. This is very useful for missions such as Measuring temperature above x km where the planet is rotating.

The commit I pushed was indeed my fault from not testing enough. You see, it works reasonably okay in Kerbin, but trying to achieve the same on Mun and other non-atmospheric bodies poses problems. (Laythe and Jool appears to be working correctly) And yes, I was too hasty and too eager to get the PR merged, which I do appologize.

The "bug" you mentioned appears because the ship is entering atmosphere. I initially thought that this mode is supposed to be used only on stable orbit, instead of re-entry, thus my lack of testing there.

To recap:

Ground Trace mode allows players to see where the craft will pass above the surface.
To use:
1. Enter a stable orbit
2. Enable Body-fixed mode, then turn on Ground Trace mode. A yellow line/curve should appear on the body, very close to the surface.
3. To get intersection with a ground target, create maneuver node and adjust the vectors until the yellow line/curve intersects with the ground target, while ensuring that the new orbit is still stable (not a collision, escape, nor entering atmosphere)
4. The yellow path underneath the vessel will move as the vessel orbits, until eventually the vessel is above the target.
5. Here, we may complete objective set by contracts such as temperature scan over site X above X km easily, from orbit.

I hope this clears things out.

**Comment by PiezPiedPy** [comment]
So as I see it the ground trace starts from the loaded craft and ends at the target?

**Comment by Charon77** [comment]
Here's the quicksave to try: [ground trace.zip](https://github.com/neuoy/KSPTrajectories/files/1105350/ground.trace.zip)

Here's the (updated) use case

In this image, we aim to get above Area F-BNVP to complete contract for temperature scan above it.
image

We would move the camera for a closer look, Focus View on Mun, and set SAS to Normal (because the target is to our North) and ensure Ground Trace mode is activated
image

We now fire our engine until the ground trace intersects the target.
Note that the ground trace is drawn 10km above surface to prevent being occluded by terrain
image

Eyeballing on Area F-BNVP shows that the ground trace intersects the area nicely
image

Fast forward a bit, and we have entered Area F-BNVP.
image

**Comment by neuoy** [comment]
Couldn't you just enable body fixed mode, and rotate the camera so that the trajectory goes through the planet center, and then you can see if it is above your target or not ? I'm not very convinced of the need for an additional display mode. (but do note that I'm in no way opposed to it).

As a side note, I think it would be more useful if the line was really following the ground, but I know this is difficult to get right (had the same problem with the impact cross, you can take a look at that code, even if it's certainly not perfect).

**Comment by fat-lobyte** [comment]
> Couldn't you just enable body fixed mode, and rotate the camera so that the trajectory goes through the planet center, and then you can see if it is above your target or not ?

I have to say that I do it in the same way, and also don't see much advantage

> Note that the ground trace is drawn 10km above surface to prevent being occluded by terrain

TBH I'm a bit reluctant to merge the feature in this state, because like that is is really not much advantage over just aligning the camera in the orbital plane.
If you would find a way to draw it directly onto the Terrain and make it visible for atmospheric patches - now that would be a game changer.

Also remember that there is no rush. I'll be around for a bit, so you can take your time, flesh out this feature and request inclusion at a later date. You might want to get community input on the Forum as well!

**Comment by Charon77** [comment]
@neuoy Would you please elaborate this:
>  I think it would be more useful if the line was really following the ground
What does it mean by following the ground?

@fat-lobyte 
>TBH I'm a bit reluctant to merge the feature in this state
That is fine with me, and I do realize this PR hadn't added significant benefits as of now.
Perhaps I would include closest approach-style marker or something like that, to help players navigate.

>If you would find a way to draw it directly onto the Terrain and make it visible for atmospheric patches - now that would be a game changer.
I will look up to that.

Thank you very much for the inputs
I really appreciate it, it certainly drives me to do a better job making PR next time 😄 
Then I believe I will post again once I've made some progress. I will ask around at the forum later on, see what they think.

**Comment by neuoy** [comment]
I mean that the line could follow the ground height, going up when there is a mountain etc, instead of being 10km above sea level. Then you could really have an increased precision compared to the camera alignment trick.

**Comment by fat-lobyte** [comment]
> Thank you very much for the inputs
> I really appreciate it, it certainly drives me to do a better job making PR next time :smile:
Then I believe I will post again once I've made some progress. I will ask around at the forum later on, see what they think.

Thanks a lot for your efforts! Looking forward to it :)

Just as a technical side note: it's better to develop features not on the master branch, but a named feature branch (like "groundtrace") or something like that. "master" should be reserved for the current state of the code. In GitHub, you have the option to allow maintainers to add commits to your Pull Request (which you apparently did). That's usually a good thing, but I don't want to be writing around on your master branch ;)

**Comment by fat-lobyte** [comment]
Hello @Charon77 ,

Based on the points discussed above, I think it's best if we close this pull request for now. You are of course very welcome to flesh-out your feature, and create a new Pull Request!

We have a new piece of documentation that helps contributors like you!
You can find it here: https://github.com/neuoy/KSPTrajectories/blob/master/CONTRIBUTING.md

I hope it provides a bit of guidance :)

Thank you for your efforts!
